File: Windows/BotSpeakerCli/Recall.cs
using System.Text;
using BotSpeaker.Cli;
using System.Text.Json.Nodes;

static partial class Cli
{
    private static string RecallHelp => $"""
        USAGE: {Name} recall <action> [options]
          configure [--region REGION] [--key-stdin]  env key or hidden prompt
          list                                      list bots and status
          status                                    configuration and local jobs
          add MEETING_URL [--name NAME]              join a meeting
          remove BOT_ID                             leave and cancel pending jobs
          speak BOT_ID TEXT [--voice ID] [--file PATH]
            [--at now|+SECONDS|ISO8601] [--loop | --repeat N] [--interval SECONDS]
          cancel JOB_ID                             stop future dispatches
          list --meeting URL_OR_ID                  filter bots to one meeting
          add ID --passcode CODE                    join Teams using ID and passcode
          add --file PATH|-                         read a full invitation
          remove-all --meeting URL_OR_ID            remove all bots in that meeting
          prepare BOT_ID TEXT [--voice ID] [--file PATH] [--wait]
          dispatch JOB_ID [--wait]                   play a prepared clip
          wait JOB_ID [--timeout SECONDS]            wait for dispatch completion
          meeting-create --file PLAN.json|-         plan turns using botId, voice, text
          meeting-start PLAN_ID [--wait]             prepare all turns, then play
          meeting-status PLAN_ID                    inspect progress
          meeting-skip PLAN_ID                       skip the current turn
          meeting-stop PLAN_ID                       stop and cancel pending turns
          meeting-wait PLAN_ID [--timeout SECONDS]   wait for the plan to finish
        speak, prepare, dispatch, and meeting-start accept --wait and --timeout.
        Plan IDs and job IDs last for this app process; stopping leaves bots joined.
        Keep the app running and awake. Times specify dispatch, not audible playback.
        Cancel cannot retract audio already sent. Output is JSON (also accepts --json).
        """;

    private static async Task<int> RecallAsync(Args args)
    {
        var action = args.Positional.FirstOrDefault() ?? "status";
        var body = new JsonObject();
        if (action == "configure")
        {
            if (args.Positional.Count != 1 || args.Options.Keys.Any(key => key is not ("json" or "region" or "key-stdin")))
                throw new UsageException("configure accepts --region, --key-stdin, and --json.");
            var key = Environment.GetEnvironmentVariable("RECALL_API_KEY") ?? Environment.GetEnvironmentVariable("RECALL_API_KEY", EnvironmentVariableTarget.User)
                ?? Environment.GetEnvironmentVariable("RECALL_AI_API_KEY") ?? Environment.GetEnvironmentVariable("RECALL_AI_API_KEY", EnvironmentVariableTarget.User);
            if (args.Flag("key-stdin")) key = ReadStdin().Trim();
            if (string.IsNullOrWhiteSpace(key) && !Console.IsInputRedirected)
            {
                Console.Error.Write("Recall API key (blank keeps saved key): ");
                var secret = new StringBuilder();
                while (true) { var k = Console.ReadKey(true); if (k.Key == ConsoleKey.Enter) break; if (k.Key == ConsoleKey.Backspace) { if (secret.Length > 0) secret.Length--; } else if (!char.IsControl(k.KeyChar)) secret.Append(k.KeyChar); }
                Console.Error.WriteLine(); key = secret.ToString();
            }
            if (!string.IsNullOrWhiteSpace(key)) body["apiKey"] = key;
            if (args.Option("region") is string region) body["region"] = region;
        }
        else
        {
            try { (action, body) = RecallCommands.Build(args.Positional, args.Options, file => file == "-" ? ReadStdin() : File.ReadAllText(file)); }
            catch (ArgumentException error) { throw new UsageException(error.Message); }
        }
        var timeout = args.Timeout;
        if (!double.IsFinite(timeout)) throw new UsageException("--timeout must be finite.");
        var client = await ControlClient.LocateAsync();
        Task<JsonObject> Request(string verb, JsonObject input) => client.PostAsync("/v1/recall/" + verb, input);
        JsonObject response;
        if (action is "wait" or "meeting-wait")
            response = await RecallCommands.WaitAsync(body["id"]!.GetValue<string>(), action == "meeting-wait", false, timeout, Request);
        else
        {
            response = await Request(action, body);
            if (args.Wait) {
                bool meeting = action == "meeting-start";
                var id = response[meeting ? "meeting" : "job"]?["id"]?.GetValue<string>() ?? throw new InvalidOperationException("Recall did not return an ID.");
                response = await RecallCommands.WaitAsync(id, meeting, action == "prepare", timeout, Request);
            }
        }
        Output.Json(response);
        return response["ok"]?.GetValue<bool>() == false ? 1 : 0;
    }
}

File: cli/Sources/botspeaker/Recall.swift
import ArgumentParser
import Foundation
import Darwin

struct Recall: AsyncParsableCommand {
    static let configuration = CommandConfiguration(abstract: "Control Recall meeting bots and schedule prerecorded speech.", discussion: "Actions: configure, status, list, add, remove, remove-all, speak, prepare, dispatch, cancel, wait, meeting-create, meeting-start, meeting-status, meeting-skip, meeting-stop, meeting-wait. Use --meeting for scoped list/removal, --file for invitations or JSON plans, and --wait for speech or meeting completion. Keep the app running and awake. Scheduling controls dispatch, not exact playback. Cancellation stops future dispatches only.")
    @OptionGroup var global: GlobalOptions
    @Argument var action: String = "status"
    @Argument var arguments: [String] = []
    @Option var region: String?
    @Option var name: String = "BotSpeaker"
    @Option(help: "ElevenLabs voice ID; defaults to the app's selected voice.") var voice: String?
    @Option(help: "now, +SECONDS, or ISO 8601 timestamp with timezone.") var at: String?
    @Option(help: "Extra gap between clips in seconds.") var interval: Double?
    @Option(name: [.customLong("repeat"), .customShort("r")]) var repeatCount: Int?
    @Flag(name: [.long, .customShort("l")]) var loop = false
    @Flag(help: "Read the Recall key from stdin rather than a hidden prompt.") var keyStdin = false
    @Option(name: [.short, .long]) var file: String?

    @Option(help: "Meeting URL or ID for list and remove-all.") var meeting: String?
    @Option(help: "Teams passcode when adding by meeting ID.") var passcode: String?
    @Flag(name: [.short, .long], help: "Wait for speak, prepare, dispatch, or meeting-start.") var wait = false
    @Option(help: "Maximum wait in seconds; work continues after timeout.") var timeout: Double = 3600

    func validate() throws {
        guard ["configure", "status", "list", "add", "remove", "remove-all", "speak", "prepare", "dispatch", "cancel", "wait", "meeting-create", "meeting-start", "meeting-status", "meeting-skip", "meeting-stop", "meeting-wait"].contains(action) else { throw ValidationError("Unknown Recall action.") }
        if loop && repeatCount != nil { throw ValidationError("Use --loop or --repeat, not both.") }
        if let repeatCount, repeatCount < 1 { throw ValidationError("--repeat must be positive.") }
        if let interval, !interval.isFinite || interval < 0 || interval > 86400 { throw ValidationError("--interval must be finite and between 0 and 86400 seconds.") }
        guard timeout.isFinite, timeout > 0 else { throw ValidationError("--timeout must be finite and positive.") }
        if wait && !["speak", "prepare", "dispatch", "meeting-start"].contains(action) { throw ValidationError("--wait is only supported for speak, prepare, dispatch, and meeting-start.") }
        if action != "speak" && (at != nil || loop || repeatCount != nil || interval != nil) { throw ValidationError("Scheduling and repeat options require speak.") }
        if file != nil && !["add", "speak", "prepare", "meeting-create"].contains(action) { throw ValidationError("--file requires add, speak, prepare, or meeting-create.") }
        if meeting != nil && !["list", "remove-all"].contains(action) { throw ValidationError("--meeting requires list or remove-all.") }
        if passcode != nil && action != "add" { throw ValidationError("--passcode requires add.") }
        if voice != nil && !["speak", "prepare"].contains(action) { throw ValidationError("--voice requires speak or prepare.") }
        if keyStdin && action != "configure" { throw ValidationError("--key-stdin requires configure.") }
        let needsID = ["remove", "speak", "prepare", "dispatch", "cancel", "wait", "meeting-start", "meeting-status", "meeting-skip", "meeting-stop", "meeting-wait"].contains(action)
        if needsID && arguments.isEmpty { throw ValidationError("An ID is required.") }
        if action == "add" && file == nil && arguments.isEmpty { throw ValidationError("A meeting invitation, URL, or ID is required.") }
        if action == "remove-all" && (meeting ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { throw ValidationError("--meeting is required for remove-all.") }
        if action == "meeting-create" && file == nil { throw ValidationError("--file is required for meeting-create.") }
        let maximum = ["speak", "prepare"].contains(action) ? Int.max : (needsID || action == "add" ? 1 : 0)
        if arguments.count > maximum { throw ValidationError("Unexpected positional arguments.") }
    }
    func run() async throws {
        do {
            let client = try await ControlClient.locate()
            let env = ProcessInfo.processInfo.environment
            var body: [String: Any] = [:]
            if action == "configure" {
                var key = env["RECALL_API_KEY"] ?? env["RECALL_AI_API_KEY"]
                if keyStdin { key = String(data: FileHandle.standardInput.readDataToEndOfFile(), encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) }
                else if key == nil, isatty(STDIN_FILENO) != 0, let entered = getpass("Recall API key (blank keeps saved key): ") { key = String(cString: entered) }
                if let key, !key.isEmpty { body["apiKey"] = key }
                if let region { body["region"] = region }
            } else if let key = env["RECALL_API_KEY"] ?? env["RECALL_AI_API_KEY"] {
                let status = try await client.post("/v1/recall/status")
                if status["configured"] as? Bool != true {
                    var config: [String: Any] = ["apiKey": key]
                    if let region { config["region"] = region }
                    _ = try await client.post("/v1/recall/configure", body: config)
                }
            }
            func read(_ path: String) throws -> String {
                path == "-" ? String(data: FileHandle.standardInput.readDataToEndOfFile(), encoding: .utf8) ?? "" : try String(contentsOfFile: path, encoding: .utf8)
            }
            if let meeting { body["meetingId"] = meeting }
            if action == "add" {
                body["meetingUrl"] = try file.map(read) ?? arguments.first
                body["name"] = name
                if let passcode { body["passcode"] = passcode }
            }
            if ["remove", "speak", "prepare"].contains(action) { body["botId"] = arguments.first }
            if ["cancel", "dispatch", "wait", "meeting-start", "meeting-status", "meeting-skip", "meeting-stop", "meeting-wait"].contains(action) { body["id"] = arguments.first }
            if action == "meeting-create" {
                guard let data = try read(file!).data(using: .utf8),
                      let plan = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw ValidationError("Meeting plan must be a JSON object.") }
                body = plan
            }
            if action == "speak" || action == "prepare" {
                let text = try file.map(read) ?? arguments.dropFirst().joined(separator: " ")
                guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { throw ValidationError("Speech text is required.") }
                body["text"] = text
                if let voice { body["voice"] = voice }
                if action == "speak" {
                    body["at"] = at ?? "now"; body["loop"] = loop; body["interval"] = interval ?? 0
                    if let repeatCount { body["repeat"] = repeatCount }
                }
            }
            var response: [String: Any]
            if action == "wait" || action == "meeting-wait" {
                response = try await waitFor(arguments[0], meeting: action == "meeting-wait", prepared: false, client: client)
            } else {
                response = try await client.post("/v1/recall/" + action, body: body)
                if wait {
                    let isMeeting = action == "meeting-start"
                    guard let id = (response[isMeeting ? "meeting" : "job"] as? [String: Any])?["id"] as? String else { throw ValidationError("Recall did not return an ID.") }
                    response = try await waitFor(id, meeting: isMeeting, prepared: action == "prepare", client: client)
                }
            }
            Output.json(response)
            if response["ok"] as? Bool == false { throw ExitCode(1) }
        } catch let exit as ExitCode { throw exit }
        catch let error as ValidationError {
            Output.fail(ControlClient.Failure(exitCode: 1, code: "recall_error", message: error.message), json: global.json)
        }
        catch { Output.fail(error, json: global.json) }
    }

    private func waitFor(_ id: String, meeting: Bool, prepared: Bool, client: ControlClient) async throws -> [String: Any] {
        let deadline = ContinuousClock.now.advanced(by: .seconds(timeout))
        while ContinuousClock.now < deadline {
            let remaining = ContinuousClock.now.duration(to: deadline)
            let seconds = Double(remaining.components.seconds) + Double(remaining.components.attoseconds) / 1e18
            let response = try await client.post("/v1/recall/" + (meeting ? "meeting-status" : "status"), body: ["id": id], timeout: max(0.001, seconds))
            let item = meeting ? response["meeting"] as? [String: Any] : (response["jobs"] as? [[String: Any]])?.first { $0["id"] as? String == id }
            guard let item else { throw ValidationError("Unknown Recall job or meeting.") }
            let state = item["status"] as? String ?? ""
            if ["failed", "cancelled", "stopped"].contains(state) { throw ValidationError(item["error"] as? String ?? "Recall work was \(state).") }
            if state == (meeting ? "finished" : prepared ? "prepared" : "finished_dispatching") { return ["ok": true, meeting ? "meeting" : "job": item] }
            try await Task.sleep(for: .milliseconds(100))
        }
        throw ValidationError("Timed out waiting for Recall. Work continues in the app; use cancel or meeting-stop to stop it.")
    }
}
